src/Tools/BuildValidator/Program.cs

```diff
@@ -226,39 +235,99 @@ private static bool ValidateFiles(IEnumerable<FileInfo> originalBinaries, BuildC
+                var compilationDiff = CompilationDiff.Create(originalBinary, optionsReader, compilation, logger, options);
                 return compilationDiff;
 
-                IMethodSymbol? getDebugEntryPoint()
+                void logResolvedMetadataReferences()
                 {
-                    if (optionsReader.GetMainTypeName() is { } mainTypeName &&
-                        optionsReader.GetMainMethodName() is { } mainMethodName)
+                    using var _ = logger.BeginScope("Metadata References");
+                    for (var i = 0; i < metadataReferenceInfos.Length; i++)
                     {
-                        var typeSymbol = compilation.GetTypeByMetadataName(mainTypeName);
-                        if (typeSymbol is object)
-                        {
-                            var methodSymbols = typeSymbol
-                                .GetMembers(mainMethodName)
-                                .OfType<IMethodSymbol>();
-                            return methodSymbols.FirstOrDefault();
-                        }
+                        logger.LogInformation($@"""{metadataReferences[i].Display}"" - {metadataReferenceInfos[i].Mvid}");
                     }
+                }
 
-                    return null;
+                void logResolvedSources()
+                {
+                    using var _ = logger.BeginScope("Source Names");
+                    foreach (var resolvedSource in sources)
+                    {
+                        var sourceFileInfo = resolvedSource.SourceFileInfo;
+                        var hash = BitConverter.ToString(sourceFileInfo.Hash).Replace("-", "");
+                        var embeddedCompressedHash = sourceFileInfo.EmbeddedCompressedHash is { } compressedHash
+                            ? ("[uncompressed]" + BitConverter.ToString(compressedHash).Replace("-", ""))
+                            : null;
+                        logger.LogInformation($@"""{resolvedSource.DisplayPath}"" - {sourceFileInfo.HashAlgorithm} - {hash} - {embeddedCompressedHash}");
+                    }
                 }
             }
             finally
             {
                 pdbReaderProvider?.Dispose();
             }
         }
+
+        private static ImmutableArray<SourceLink> ResolveSourceLinks(CompilationOptionsReader compilationOptionsReader, ILogger logger)
+        {
+            using var _ = logger.BeginScope("Source Links");
+
+            var sourceLinkUTF8 = compilationOptionsReader.GetSourceLinkUTF8();
+            if (sourceLinkUTF8 is null)
+            {
+                return default;
+            }
+
+            var parseResult = JsonConvert.DeserializeAnonymousType(Encoding.UTF8.GetString(sourceLinkUTF8), new { documents = (Dictionary<string, string>?)null });
+            var sourceLinks = parseResult.documents.Select(makeSourceLink).ToImmutableArray();
+
+            if (sourceLinks.IsDefault)
```

Comment:
>sourceLinks.IsDefault [](start = 16, length = 21)

Does `ToImmutableArray()` ever return `default`?

Reply:
I do not think so.